Exposing ConvexMonotoneInterpolation

Hi QL Team,

Exposing `LinearFlatInterpolation` in `SWIG/interpolation.i` was pretty straightforward, but we are running into `python setup.py build` issues with `ConvexMonotoneInterpolation` because it is a template class and cannot be passed to `make_safe_interpolation`:

```QuantLib/quantlib_wrap.cpp:9215:64: error: type/value mismatch at argument 1 in template parameter list for `‘template<class I> class SafeInterpolation`’
 typedef SafeInterpolation<QuantLib::ConvexMonotoneInterpolation> SafeConvexMonotoneInterpolation;```

Does anyone know what additional changes should happen to expose `ConvexMonotoneInterpolation`? We have limited C++/Swig experience so apologies if it is obvious.

It looks like someone had made an attempt a few years ago but there was no follow-up:

https://sourceforge.net/p/quantlib/mailman/message/32391881/

Thank you!

-Miguel

--- a/SWIG/interpolation.i
+++ b/SWIG/interpolation.i
@@ -228,3 +228,23 @@ class RichardsonExtrapolation {
 };
 
 #endif
+
+
+%{
+namespace QuantLib {
+    typedef ConvexMonotoneInterpolation<Array::const_iterator, Array::const_iterator> 
+        ConvexMonotoneArrayInterpolation;
+}
+
+typedef SafeInterpolation<QuantLib::ConvexMonotoneArrayInterpolation> SafeConvexMonotoneInterpolation;
+%}
+
+%rename(ConvexMonotoneInterpolation) SafeConvexMonotoneInterpolation;
+class SafeConvexMonotoneInterpolation {
+    #if defined(SWIGCSHARP)
+    %rename(call) operator();
+    #endif
+  public:
+    SafeConvexMonotoneInterpolation(const Array& x, const Array& y);
+    Real operator()(Real x, bool allowExtrapolation=false);
+};

--- a/SWIG/interpolation.i
+++ b/SWIG/interpolation.i
@@ -231,12 +231,20 @@ class RichardsonExtrapolation {
 
 
 %{
-namespace QuantLib {
-    typedef ConvexMonotoneInterpolation<Array::const_iterator, Array::const_iterator> 
-        ConvexMonotoneArrayInterpolation;
-}
-
-typedef SafeInterpolation<QuantLib::ConvexMonotoneArrayInterpolation> SafeConvexMonotoneInterpolation;
+class SafeConvexMonotoneInterpolation {
+  public:
+    SafeConvexMonotoneInterpolation(const Array& x, const Array& y,
+                                    Real quadraticity = 0.3,
+                                    Real monotonicity = 0.7,
+                                    bool forcePositive = true)
+    : x_(x), y_(y), f_(x_.begin(), x_.end(), y_.begin(),
+                       quadraticity, monotonicity, forcePositive) {}
+    Real operator()(Real x, bool allowExtrapolation=false) {
+        return f_(x, allowExtrapolation);
+    }
+    Array x_, y_;
+    QuantLib::ConvexMonotoneInterpolation<Array::const_iterator, Array::const_iterator> f_;
+};
 %}
 
 %rename(ConvexMonotoneInterpolation) SafeConvexMonotoneInterpolation;
@@ -245,6 +253,10 @@ class SafeConvexMonotoneInterpolation {
     %rename(call) operator();
     #endif
   public:
-    SafeConvexMonotoneInterpolation(const Array& x, const Array& y);
+    SafeConvexMonotoneInterpolation(const Array& x, const Array& y,
+                                    Real quadraticity = 0.3,
+                                    Real monotonicity = 0.7,
+                                    bool forcePositive = true);
     Real operator()(Real x, bool allowExtrapolation=false);
 };
+

